### ISSUE TITLE
bug: IonModal does not close properly when close is called in catch of async function

### ISSUE THREAD

**Issue by ldikmans** [ionic-team/ionic-framework#22336]
# Bug Report

**Ionic version:**

[] **4.x**
[x] **5.x**

**Current behavior:**

When a modal is closed in a catch of a callback function, the content of the modal is removed but the modal stays on the page as a white screen overlaying everything.

**Expected behavior:**

The modal should disappear

**Steps to reproduce:**

See sample application on github

1. Create an App.tsx with IonRouterOutlet that contains 3 entries: the redirect to home from /, home and modalPage
2. Create a home page with a button to navigate to the modal page
3. The modal page contains a modal with a component (Posts) that calls a API. Pass the setShowModal as a function to this component and a boolean to force an error in the API.
4. Open the modal with the "Open Modal" button and close it by pressing the Close Me button at the bottom of the list. Normal behaviour occurs: the modal closes. 
5. Open the modal again, but force an error in the async function that calls the API. When the error is caught, the modal is removed but an overlay appears that hides the page. 

Note: when I try the same setup without the RouterOutlet (only one page, ModalPage) the error does not occur.

**Related code:**

[https://github.com/ldikmans/ionic-react-modal-bug](https://github.com/ldikmans/ionic-react-modal-bug)

**Note**: I also have the example on stackblitz [https://stackblitz.com/edit/react-ts-k6egca](https://stackblitz.com/edit/react-ts-k6egca?devtoolsheight=33&file=src/index.tsx). However, the router shows some strange behaviour in the iFrame. To reproduce there, you need to refresh the page to make the buttons on the modalPage show up.

**Note**: On stackblitz the second button ("Open Modal that throws and error and closes", see screenshot) needs to be pressed multiple times to reproduce the issue. 

**Note**. Getting the error can be forced in multiple ways, for example calling an API that requires authorization. The boolean is just a way to force this

```
The call back looks as follows:

`  const getPosts = useCallback(async () => {
        try {
            const userPosts = await axios.get("https://jsonplaceholder.typicode.com/posts")
            setPosts(userPosts.data);
            if (error) {
                throw new Error('error thrown');
            }

        } catch (err) {
            console.error(err.message);
            setShowModal(false);
        }
    }, [setShowModal, error]);

    useEffect(() => {
        getPosts();
    }, [getPosts]);
`

```

**Other information:**

I think that maybe what I report here is a special case of this bug, however I am not sure about that
[https://github.com/ionic-team/ionic-framework/issues/19377](https://github.com/ionic-team/ionic-framework/issues/19377)

To reproduce the issue run the sample application. Click on the first button to open the Modal and close it. Then click on the second button. Screenshot of issue: 
![image](https://user-images.githubusercontent.com/14298582/96297154-b816c600-0ff0-11eb-8ef5-03d09a63e186.png)

**Ionic info:**

```
Ionic:

   Ionic CLI       : 6.11.7 (/usr/local/lib/node_modules/@ionic/cli)
   Ionic Framework : @ionic/react 5.4.0

Capacitor:

   Capacitor CLI   : not installed
   @capacitor/core : 2.4.2

Utility:

   cordova-res                          : 0.15.1
   native-run (update available: 1.2.1) : 1.0.0

System:

   NodeJS : v12.18.3 (/usr/local/Cellar/node@12/12.18.3/bin/node)
   npm    : 6.14.6
   OS     : macOS Catalina
```

**Comment by ldikmans** [comment]
Did anybody get a chance to take a look at this issue?

We have a workaround in place to remove modals that are visible:
 ```
useIonViewWillEnter(() => {
        //see bugreport https://github.com/ionic-team/ionic-framework/issues/22336
        const ionOverlay = 'ion-overlay-';
        for (let i = 1; i < 100; i++) {
            let overlay = ionOverlay.concat(i.toString());
            const modal = document.getElementById(overlay) as HTMLIonInputElement;
            if (modal) {
                const parent = modal.parentNode;
                parent?.removeChild(modal);
                break;
            }
        }
    });
```
Obviously, we would like to get rid of this as soon as possible, for many reasons

**Comment by rluttikhuizen** [comment]
Hello,

Any update on this issue? I have a similar problem.

Thanks,

Ronald

**Comment by ldikmans** [comment]
can someone look at this, I am more than willing to help out. @elylucas is there any way I can help to solve this issue?

**Comment by elylucas** [comment]
Hi @ldikmans,

I took a look at your repo, and I think the issue is with the `<IonContent>` container in the Posts.tsx file. You shouldn't need that and I think its confusing the framework on trying to remove all the elements when closing the modal. When I took it out locally it seemed to work properly.

Going to close this issue for now, but ping me if you have more questions.

Thanks

**Comment by ldikmans** [comment]
Hi @elylucas 

Thanks for the response! Unfortunately this does not improve the situation but it makes it worse.  I can no longer scroll in the modal and the rest of the behaviour is the same. 
Can you please reopen or explain in more detail what you changed to fix the behavior? 

I  changed the code in Posts.tsx to the following (the Fragment is needed otherwise it won't compile): 

`
    return (
        <Fragment>
            <IonItem className="ion-no-padding ion-no-margin" lines="full">
                <div tabIndex={-1} />
                <IonLabel className="ion-no-padding ion-no-margin" >
                    <div className="ion-text-start ion-text-wrap">Posts​​​​​</div>
                </IonLabel>
            </IonItem>
            <IonList>
                {posts?.map((post) => (
                    <IonItem key={post.id}>
                        <IonLabel>{post.title}
                        </IonLabel>
                    </IonItem>
                ))}
                <IonButton onClick={() => { closeModal() }}>Close me</IonButton>
            </IonList>
        </Fragment>
    );
};`

**Comment by elylucas** [comment]
@ldikmans 

Ok, keeping the IonContent in Posts.tsx, try setting the hideModal function in ModalPage.tsx, try setting throwError to false like so and let me know how it goes:

```  
function hideModal() {
    console.log('hiding the modal');
    setThrowError(false);
  }
```

**Comment by ldikmans** [comment]
Thank you for the quick response. 

I updated the code as per your suggestion, I updated the function hideModal with a 'setThrowError(false)'. There is no difference in behavior, the problem still occurs. 

Also note that in our real app we don't have a state for throwError, I only created that to simulate our case. In the real situation, the API call to our backend returns a 401 and as a response we navigate to the login page. This seems to be the only time we have this issue: if an error occurs in the catch of the call with axios.
